### PR TITLE
Add avatar support for codeberg in git blame

### PR DESCRIPTION
**Pull request by WeetHet**
Release Notes:

- Added support for avatars in git blame for repositories hosted on codeberg

Screenshot 2024-04-25 at 16 45 22

Questions:
- Should we move git stuff like `Commit`, `Author`, etc outside of hosting-specific files (I don't think so, as other hostings can have different stuff)
- Should we also add support for self hosted forgejo instances or should it be a different PR?

### CODE REVIEW

**Comment by WeetHet** [comment]
I've converted this to a draft as some things have to be decided on first

File: crates/git/src/hosting_provider.rs [hunk 1]
```diff
@@ -71,24 +71,25 @@ impl HostingProvider {
...
-                    let mut url = Url::parse(&author.avatar_url)?;
-                    url.set_query(Some("size=128"));
-                    Some(url)
-                } else {
-                    None
-                };
+        Ok(author
+            .map(|author| -> Result<Url, url::ParseError> {
+                let mut url = Url::parse(&author.avatar_url)?;
+                url.set_query(Some("size=128"));
```
Comment (mrnugget) [inline crates/git/src/hosting_provider.rs:89]: Does codeberg even support this?

Comment (WeetHet) [reply crates/git/src/hosting_provider.rs:89]: Size? Probably not, but it doesn't break it, so whatever I guess?

Comment (WeetHet) [reply crates/git/src/hosting_provider.rs:89]: I can add a match inside to only set it for GitHub

**Comment by WeetHet** [comment]
> > * Should we move git stuff like `Commit`, `Author`, etc outside of hosting-specific files (I don't think so, as other hostings can have different stuff)
> 
> No, I don't think that's necessary. These are really just there to match GitHub's API schema.
> 
> I think it's fine for now for all the providers to simply return a URL.

I've made it a struct holding only the URL for now, if in the future other stuff will be needed, we can always add new fields

File: Cargo.lock [hunk 2]
```diff
@@ -11095,6 +11095,7 @@ dependencies = [
  "futures 0.3.28",
  "gpui",
  "indoc",
+ "itertools 0.11.0",
```
Comment (mrnugget) [inline Cargo.lock:11098]: I don't think you need this. It's fixed on `main`.

File: crates/git/src/hosting_provider.rs [hunk 3]
```diff
@@ -71,24 +71,32 @@ impl HostingProvider {
...
+            }
+            HostingProvider::Codeberg => {
+                let commit = commit.to_string();
+                (
+                    codeberg::fetch_codeberg_commit_author(repo_owner, repo, &commit, &client)
+                        .await?,
+                    None,
+                )
+            }
+            _ => (None, None),
```
Comment (mrnugget) [inline crates/git/src/hosting_provider.rs:90]: This seems overly complicated: when only 1/3 cases uses the URL query, I think we should set the size inside the `HostingProvider::Github` arm.

Comment (WeetHet) [reply crates/git/src/hosting_provider.rs]: Okay, done

File: Cargo.lock (hunk 2 above)
Comment (WeetHet) [reply Cargo.lock]: Cargo really insists that I do need this, but okay, I've forcefully removed it for now

**Review by mrnugget — APPROVED** [review]

**Pull request merged**